// ngx-sharebuttons/esm2020/popup/share-buttons-popup.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { SHARE_BUTTONS_CONFIG } from 'ngx-sharebuttons';
import { ShareButtonsModule } from 'ngx-sharebuttons/buttons';
import { ShareButtonsPopup } from './share-buttons-popup';
import { SharePopupService } from './share-popup.service';
import { SharePopupButtonsDirective } from './share-popup-buttons.directive';
import { SharePopupComponent } from './share-popup.component';
import * as i0 from "@angular/core";
/**
 * @deprecated The pop-up share buttons module is deprecated since version 12.0.0 and will be removed in v13.0.0
 */
export class ShareButtonsPopupModule {
    static withConfig(config) {
        return {
            ngModule: ShareButtonsModule,
            providers: [{ provide: SHARE_BUTTONS_CONFIG, useValue: config }]
        };
    }
}
ShareButtonsPopupModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.4", ngImport: i0, type: ShareButtonsPopupModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ShareButtonsPopupModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.4", ngImport: i0, type: ShareButtonsPopupModule, declarations: [ShareButtonsPopup,
        SharePopupButtonsDirective,
        SharePopupComponent], imports: [CommonModule,
        ShareButtonsModule], exports: [ShareButtonsModule,
        ShareButtonsPopup,
        SharePopupButtonsDirective,
        SharePopupComponent] });
ShareButtonsPopupModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.4", ngImport: i0, type: ShareButtonsPopupModule, providers: [
        SharePopupService
    ], imports: [CommonModule,
        ShareButtonsModule, ShareButtonsModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.4", ngImport: i0, type: ShareButtonsPopupModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        ShareButtonsPopup,
                        SharePopupButtonsDirective,
                        SharePopupComponent
                    ],
                    imports: [
                        CommonModule,
                        ShareButtonsModule
                    ],
                    exports: [
                        ShareButtonsModule,
                        ShareButtonsPopup,
                        SharePopupButtonsDirective,
                        SharePopupComponent
                    ],
                    providers: [
                        SharePopupService
                    ]
                }]
        }] });
