// ngx-hm-carousel/esm2020/lib/ngx-hm-carousel.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { NgxHmCarouselDynamicDirective } from './ngx-hm-carousel-dynamic.directive';
import { NgxHmCarouselItemDirective } from './ngx-hm-carousel-item.directive';
import { NgxHmCarouselComponent } from './ngx-hm-carousel.component';
import * as i0 from "@angular/core";
export class NgxHmCarouselModule {
}
NgxHmCarouselModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.0.1", ngImport: i0, type: NgxHmCarouselModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgxHmCarouselModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.0.1", ngImport: i0, type: NgxHmCarouselModule, declarations: [NgxHmCarouselComponent,
        NgxHmCarouselDynamicDirective,
        NgxHmCarouselItemDirective], imports: [CommonModule, FormsModule], exports: [NgxHmCarouselComponent,
        NgxHmCarouselDynamicDirective,
        NgxHmCarouselItemDirective] });
NgxHmCarouselModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.0.1", ngImport: i0, type: NgxHmCarouselModule, imports: [[CommonModule, FormsModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.0.1", ngImport: i0, type: NgxHmCarouselModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, FormsModule],
                    declarations: [
                        NgxHmCarouselComponent,
                        NgxHmCarouselDynamicDirective,
                        NgxHmCarouselItemDirective,
                    ],
                    exports: [
                        NgxHmCarouselComponent,
                        NgxHmCarouselDynamicDirective,
                        NgxHmCarouselItemDirective,
                    ],
                }]
        }] });
