// angular2-hotkeys/esm2020/lib/hotkey.module.mjs
import { NgModule } from '@angular/core';
import { HotkeysDirective } from './hotkeys.directive';
import { HotkeysCheatsheetComponent } from './hotkeys-cheatsheet/hotkeys-cheatsheet.component';
import { CommonModule } from '@angular/common';
import { HotkeyOptions } from './hotkey.options';
import { HotkeysService } from './hotkeys.service';
import * as i0 from "@angular/core";
export class HotkeyModule {
    // noinspection JSUnusedGlobalSymbols
    static forRoot(options = {}) {
        return {
            ngModule: HotkeyModule,
            providers: [
                HotkeysService,
                { provide: HotkeyOptions, useValue: options }
            ]
        };
    }
}
HotkeyModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.7", ngImport: i0, type: HotkeyModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
HotkeyModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.7", ngImport: i0, type: HotkeyModule, declarations: [HotkeysDirective, HotkeysCheatsheetComponent], imports: [CommonModule], exports: [HotkeysDirective, HotkeysCheatsheetComponent] });
HotkeyModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.7", ngImport: i0, type: HotkeyModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.7", ngImport: i0, type: HotkeyModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [HotkeysDirective, HotkeysCheatsheetComponent],
                    imports: [CommonModule],
                    exports: [HotkeysDirective, HotkeysCheatsheetComponent]
                }]
        }] });
