// ngx-drag-drop/esm2020/lib/dnd.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { DndDraggableDirective, DndDragImageRefDirective, } from './dnd-draggable.directive';
import { DndDropzoneDirective, DndPlaceholderRefDirective, } from './dnd-dropzone.directive';
import { DndHandleDirective } from './dnd-handle.directive';
import * as i0 from "@angular/core";
export class DndModule {
}
DndModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.6", ngImport: i0, type: DndModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DndModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.6", ngImport: i0, type: DndModule, declarations: [DndDraggableDirective,
        DndDropzoneDirective,
        DndHandleDirective,
        DndPlaceholderRefDirective,
        DndDragImageRefDirective], imports: [CommonModule], exports: [DndDraggableDirective,
        DndDropzoneDirective,
        DndHandleDirective,
        DndPlaceholderRefDirective,
        DndDragImageRefDirective] });
DndModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.6", ngImport: i0, type: DndModule, imports: [CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.6", ngImport: i0, type: DndModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        DndDraggableDirective,
                        DndDropzoneDirective,
                        DndHandleDirective,
                        DndPlaceholderRefDirective,
                        DndDragImageRefDirective,
                    ],
                    exports: [
                        DndDraggableDirective,
                        DndDropzoneDirective,
                        DndHandleDirective,
                        DndPlaceholderRefDirective,
                        DndDragImageRefDirective,
                    ],
                    imports: [CommonModule],
                }]
        }] });
