// @flowjs/ngx-flow/esm2020/lib/ngx-flow.module.mjs
import { NgModule } from '@angular/core';
// import Flow from '@flowjs/flow.js';
import { ButtonDirective } from './button.directive';
import { DropDirective } from './drop.directive';
import { FlowDirective } from './flow.directive';
import { SrcDirective } from './src.directive';
import * as i0 from "@angular/core";
const directives = [
    ButtonDirective,
    SrcDirective,
    DropDirective,
    FlowDirective,
];
// export function flowFactory() {
//   return Flow;
// }
export class NgxFlowModule {
}
NgxFlowModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.5", ngImport: i0, type: NgxFlowModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgxFlowModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.0.5", ngImport: i0, type: NgxFlowModule, declarations: [ButtonDirective,
        SrcDirective,
        DropDirective,
        FlowDirective], exports: [ButtonDirective,
        SrcDirective,
        DropDirective,
        FlowDirective] });
NgxFlowModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.0.5", ngImport: i0, type: NgxFlowModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.5", ngImport: i0, type: NgxFlowModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [],
                    declarations: directives,
                    // providers: [
                    //   {
                    //     provide: FlowInjectionToken,
                    //     useFactory: flowFactory
                    //   }
                    // ],
                    exports: directives,
                }]
        }] });
