// @flowjs/ngx-flow/esm2020/lib/src.directive.mjs
import { Directive, Input } from '@angular/core';
import * as i0 from "@angular/core";
export class SrcDirective {
    constructor(el) {
        this.el = el;
        this.fileReader = new FileReader();
    }
    set flowSrc(transfer) {
        this.fileReader.readAsDataURL(transfer.flowFile.file);
        this.fileReader.onload = (event) => {
            const url = event.target.result;
            this.el.nativeElement.setAttribute('src', url);
        };
    }
}
SrcDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.5", ngImport: i0, type: SrcDirective, deps: [{ token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Directive });
SrcDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.0.5", type: SrcDirective, selector: "[flowSrc]", inputs: { flowSrc: "flowSrc" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.5", ngImport: i0, type: SrcDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[flowSrc]'
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }]; }, propDecorators: { flowSrc: [{
                type: Input
            }] } });
